engine/src/test/java/io/camunda/zeebe/engine/processing/timer/TimerStartEventTest.java

```diff
@@ -877,4 +877,298 @@ public void shouldTriggerOnlyTimerStartEvent() {
+                .withElementId("process_2")
+                .withProcessDefinitionKey(secondDeploymentProcessDefinitionKey)
+                .count())
+        .isEqualTo(3);
+  }
+
+  @Test
+  public void shouldAvoidTimeShifting() {
+    // given
+    // Set the start time to 10 seconds later
+    final ZonedDateTime start =
+        ZonedDateTime.ofInstant(
+                Instant.ofEpochMilli(engine.getClock().getCurrentTimeInMillis()),
+                ZoneId.systemDefault())
+            .plusSeconds(10);
+
+    final long dueDate = start.plusSeconds(10).toInstant().toEpochMilli();
+    final long lastDueDate = start.plusSeconds(20).toInstant().toEpochMilli();
+    final BpmnModelInstance model =
+        Bpmn.createExecutableProcess("process")
+            .startEvent("start")
+            .timerWithCycle(String.format("R2/%s/PT10S", start))
+            .endEvent("end")
+            .done();
+    final var deployedProcess =
+        engine
+            .deployment()
+            .withXmlResource(model)
+            .deploy()
+            .getValue()
+            .getProcessesMetadata()
+            .get(0);
+    final long processDefinitionKey = deployedProcess.getProcessDefinitionKey();
+
+    // when
+    engine.increaseTime(Duration.ofSeconds(25));
+
+    // then
+    final TimerRecordValue timerRecord =
+        RecordingExporter.timerRecords(TimerIntent.TRIGGERED)
+            .withProcessDefinitionKey(deployedProcess.getProcessDefinitionKey())
+            .getFirst()
+            .getValue();
+
+    Assertions.assertThat(timerRecord)
+        .hasDueDate(dueDate)
+        .hasTargetElementId("start")
+        .hasElementInstanceKey(TimerInstance.NO_ELEMENT_INSTANCE);
+
+    assertThat(
+            RecordingExporter.processInstanceRecords(ProcessInstanceIntent.ELEMENT_ACTIVATED)
+                .withElementId("end")
+                .withProcessDefinitionKey(deployedProcess.getProcessDefinitionKey())
+                .exists())
+        .isTrue();
+
+    // when
+    engine.increaseTime(Duration.ofSeconds(5));
+
+    // then
+    final TimerRecordValue lastTimerRecord =
+        RecordingExporter.timerRecords(TimerIntent.TRIGGERED)
+            .withProcessDefinitionKey(deployedProcess.getProcessDefinitionKey())
+            .getLast()
+            .getValue();
+
+    Assertions.assertThat(lastTimerRecord)
+        .hasDueDate(lastDueDate)
+        .hasTargetElementId("start")
+        .hasElementInstanceKey(TimerInstance.NO_ELEMENT_INSTANCE);
+
+    assertThat(
+            RecordingExporter.timerRecords(TimerIntent.TRIGGERED)
+                .withProcessDefinitionKey(processDefinitionKey)
+                .count())
+        .isEqualTo(2);
+    assertThat(
+            RecordingExporter.processInstanceRecords(ProcessInstanceIntent.ELEMENT_ACTIVATING)
+                .withElementId("process")
+                .withProcessDefinitionKey(processDefinitionKey)
+                .count())
+        .isEqualTo(2);
```

Comment:
:x: We should avoid using `.count()` without a limit. The same as in the other comment.

Instead of waiting (10 seconds), let's limit the stream to 2 records using `.limit(2)`. We expect that the timer is triggered two times.